### DISCUSSION TITLE
Tor perfclient statistics and simulation time?

### DISCUSSION THREAD

**Discussion by timwalsh300** [shadow/shadow#2653] [Q&A] [answered]
Hi, I'm a graduate student getting started in Tor research, experimenting with Shadow and tornettools over the past couple of months. I'm amazed by the community and the ecosystem of tools, and I hope to meet some of you at conferences over the next few years.

Anyway, when I run the example 60-minute, 1% scale tornettools simulation, the parse and plot commands produce the expected output. But when I've run a number of 15-minute simulations, configured for my own experiments, the charts for 50 KB, 1 MB, and 5 MB perfclient transfer times only have the blue line for "Public Tor." Is there a minimum simulation time required to generate those statistics? Or did I break something with my modifications to the configuration?

**Comment by stevenengler** [comment, marked as answer]
Hopefully Shadow and tornettools are helpful in your research! You probably need to update the value of `--converge-time` for the `tornettools parse` step. It needs to be shorter than the total simulation runtime.

```
$ tornettools parse --help
[...]
  -c N, --converge-time N
                        The number of seconds after the beginning of the
                        simulation that we should ignore in the tgen and
                        oniontrace log files, i.e., so we don't track network
                        performance before the network has reached steady-state.
                        Log messages during the interval [0, converge_time) will
                        be ignored. (default: 1200)
```

The default is 1200 seconds, which is 20 minutes. You should try 600 (10 minutes) instead. We should probably mention this in documentation somewhere.

**Reply by timwalsh300** [reply]
Thank you, that sounds right. I'll try it in the next simulation that I need to run and report back. The explanation for this option also makes me think that I need to reconsider the design of my experiment, if my simulation isn't even reaching steady-state...

**Reply by timwalsh300** [reply]
Yes, that worked perfectly. I just realized that I could try it without even needing to re-run any of my simulations, because my data is already sitting there and this is only a post-processing step.

**Reply by robgjansen** [reply]
Great! Just wanted to note that it's possible that your network has reached a good converged state before 20 minutes, but this depends a lot on your specific network. We found that 20 minutes is a good convergence time that should be safe to use in networks of almost any size.

**Comment by akhlaghali991** [comment]
Hello
I am a master student.
Can you help me to implement the Tor network? (Software shadow or cisco and ...)

**Reply by robgjansen** [reply]
You can get started using Shadow by following the guide here:
https://shadow.github.io/docs/guide/shadow.html

If you still have questions after reading the guide, please post them in a new discussion item. It will help if you can be more specific about your goals and roadblocks.

**Reply by WangxyChocolate** [reply]
> Hello I am a master student. Can you help me to implement the Tor network? (Software shadow or cisco and ...)

So what about your Tor network?Can we discuss something?
